hazelcast/src/main/java/com/hazelcast/map/impl/query/MapQueryEngineImpl.java

```diff
@@ -16,707 +16,215 @@
-import static java.util.concurrent.TimeUnit.MINUTES;
 
 /**
- * The {@link MapQueryEngine} implementation.
+ * Invokes query logic using the QueryDispatcher.
+ * Knows nothing about query logic or how to dispatch query operations. It relies on query dispatcher only.
+ * It never dispatches query operations directly.
  */
 public class MapQueryEngineImpl implements MapQueryEngine {
 
-    protected static final long QUERY_EXECUTION_TIMEOUT_MINUTES = 5;
-
-    protected final MapServiceContext mapServiceContext;
-    protected final NodeEngine nodeEngine;
-    protected final ILogger logger;
-    protected final QueryResultSizeLimiter queryResultSizeLimiter;
-    protected final InternalSerializationService serializationService;
-    protected final IPartitionService partitionService;
-    protected final QueryOptimizer queryOptimizer;
-    protected final OperationService operationService;
-    protected final ClusterService clusterService;
-    protected final LocalMapStatsProvider localMapStatsProvider;
-    protected final boolean parallelEvaluation;
-    protected final ManagedExecutorService executor;
-
-    public MapQueryEngineImpl(MapServiceContext mapServiceContext, QueryOptimizer optimizer) {
+    protected MapServiceContext mapServiceContext;
+    protected NodeEngine nodeEngine;
+    protected ILogger logger;
+    protected QueryResultSizeLimiter queryResultSizeLimiter;
+    protected InternalSerializationService serializationService;
+    protected IPartitionService partitionService;
+    protected OperationService operationService;
+    protected ClusterService clusterService;
+    protected LocalMapStatsProvider localMapStatsProvider;
+    protected ManagedExecutorService executor;
+    protected MapQueryDispatcher queryDispatcher;
+
+    public MapQueryEngineImpl(MapServiceContext mapServiceContext) {
         this.mapServiceContext = mapServiceContext;
         this.nodeEngine = mapServiceContext.getNodeEngine();
         this.serializationService = (InternalSerializationService) nodeEngine.getSerializationService();
         this.partitionService = nodeEngine.getPartitionService();
         this.logger = nodeEngine.getLogger(getClass());
         this.queryResultSizeLimiter = new QueryResultSizeLimiter(mapServiceContext, logger);
-        this.queryOptimizer = optimizer;
         this.operationService = nodeEngine.getOperationService();
         this.clusterService = nodeEngine.getClusterService();
         this.localMapStatsProvider = mapServiceContext.getLocalMapStatsProvider();
-        this.parallelEvaluation = nodeEngine.getProperties().getBoolean(QUERY_PREDICATE_PARALLEL_EVALUATION);
         this.executor = nodeEngine.getExecutionService().getExecutor(QUERY_EXECUTOR);
+        this.queryDispatcher = new MapQueryDispatcher(mapServiceContext);
     }
 
-    QueryResultSizeLimiter getQueryResultSizeLimiter() {
-        return queryResultSizeLimiter;
-    }
-
+    // invoked from proxy layer
```

Comment:
Don't get what you mean,but the purpose of this comment was to say that this class should be only used from the top-level proxy layer (like MapProxy, etc.) I've adjusted the javadoc